### PR TITLE
Prefer defaulted default constructor for Bitset

### PR DESCRIPTION
**Pull request by dalg24**
The default argument to the constructor that takes the size of the bitset was deferring to another constructor that creates an empty view with a label argument.  This alocates 128 bits for the view header. This showed when constructing an UnorderedMap with a pointless 128-bit "header-only" allocation which implies an unnecessary fence.

cc @uliegecsm @romintomasetti

### CODE REVIEW

File: containers/src/Kokkos_Bitset.hpp [hunk 1]
```diff
@@ -92,9 +92,10 @@ class Bitset {
   using block_view_type = View<unsigned*, Device, MemoryTraits<RandomAccess>>;
 
  public:
-  /// constructor
+  Bitset() = default;
```
Comment (romintomasetti) [inline containers/src/Kokkos_Bitset.hpp:95]: Just wondering... Is `KOKKOS_DEFAULTED_FUNCTION` still needed to make the bitset default constructible on device?

**Comment by romintomasetti** [comment]
> Fixes #6467.

@masterleinad I fixes part of #6467 I think. What about https://github.com/kokkos/kokkos/issues/6467#issuecomment-1749076147 ?

File: containers/src/Kokkos_Bitset.hpp (hunk 1 above)
Comment (dalg24) [reply containers/src/Kokkos_Bitset.hpp:95]: It was not constructible from device before and I see no good reason to change it here

**Comment by masterleinad** [comment]
> @masterleinad I fixes part of #6467 I think. What about [#6467 (comment)](https://github.com/kokkos/kokkos/issues/6467#issuecomment-1749076147) ?

That's a separate discussion and isn't reflected in the title or description of the pull request and only serves as a motivation. Open a separate issue for it if you want to discuss it in a larger context.

File: containers/src/Kokkos_Bitset.hpp [hunk 2]
```diff
@@ -92,9 +92,10 @@ class Bitset {
...
+
   /// arg_size := number of bit in set
-  Bitset(unsigned arg_size = 0u) : Bitset(Kokkos::view_alloc(), arg_size) {}
+  Bitset(unsigned arg_size) : Bitset(Kokkos::view_alloc(), arg_size) {}
```
[the remark this answers is not in the document]
Comment (dalg24) [reply containers/src/Kokkos_Bitset.hpp:98]: I tend to agree with the suggested change but I will note that it was not explicit before and that is besides the scope of this PR.
That said I will apply the suggested change if other reviewers would like me to.

File: containers/src/Kokkos_Bitset.hpp (hunk 1 above)
Comment (romintomasetti) [reply containers/src/Kokkos_Bitset.hpp:95]: Ok!

**Comment by romintomasetti** [comment]
@dal24 @masterleinad Would you like me to open a new issue with https://github.com/kokkos/kokkos/issues/6467#issuecomment-1749076147 ?

**Comment by dalg24** [comment]
> @dal24 @masterleinad Would you like me to open a new issue with [#6467 (comment)](https://github.com/kokkos/kokkos/issues/6467#issuecomment-1749076147) ?

Please do so

**Comment by romintomasetti** [comment]
> > @dal24 @masterleinad Would you like me to open a new issue with [#6467 (comment)](https://github.com/kokkos/kokkos/issues/6467#issuecomment-1749076147) ?
> 
> Please do so

Just did, it is #6526. Please tell me if the title is OK with you :wink:

**Comment by masterleinad** [comment]
```
/__w/kokkos/kokkos/containers/unit_tests/TestBitset.hpp:159: Failure
Value of: b1.is_allocated()
  Actual: false
Expected: true
```
which was anticipated in https://github.com/kokkos/kokkos/issues/6467#issue-1911018394.

**Review by masterleinad — APPROVED** [review]

**Comment by dalg24** [comment]
SYCL build timed out but I intend to ignore if I can get a 2nd approval

File: containers/unit_tests/TestBitset.hpp [hunk 3]
```diff
@@ -155,7 +155,7 @@ void test_bitset() {
 
   {
     unsigned ts = 100u;
-    bitset_type b1;
+    bitset_type b1(Kokkos::view_alloc("MyBitset"), 0);
```
Comment (romintomasetti) [inline containers/unit_tests/TestBitset.hpp:158]: For such tests, shouldn't `Kokkos::Test::Tools` with `Config::EnableAllocs` be used ?

Comment (dalg24) [reply containers/unit_tests/TestBitset.hpp:160]: Sure

**Review by romintomasetti — APPROVED** [review]

File: containers/unit_tests/TestBitset.hpp [hunk 4]
```diff
@@ -237,6 +242,44 @@ void test_bitset() {
...
+        Kokkos::Bitset bs1(0);
+        EXPECT_TRUE(bs1.is_allocated());
+
+        Kokkos::Bitset bs2(Kokkos::view_alloc("MyBitset"), 0);
+        EXPECT_TRUE(bs2.is_allocated());
+      },
+      [&](AllocateDataEvent) {
+        return MatchDiagnostic{true, {"Found alloc event"}};
+      });
+  ASSERT_TRUE(success);
```
Comment (romintomasetti) [inline containers/unit_tests/TestBitset.hpp:278]: Shouldn't this test be done in 2 distinct `validate_existence` ? If I understand it correctly, in its current state, the first allocation `bs1(0)` could in fact not allocate anything and the test would still pass if `bs2` allocates something?

Not requesting any change though, I just wanted to make sure I understand `validate_existence` :smile:

BTW is there any helper in the `Kokkos` testing toolbox that would not only allow us to ensure that some event occurred (in this case `AllocateDataEvent`) but also to count them ?

Comment (dalg24) [reply containers/unit_tests/TestBitset.hpp:278]: You are correct and that is how I had written the test initially but I didn't think it was worth the boiler plate code.

I don't know that there is currently something to count the number of allocations even though writing directly a tool that does that would be trivial.

**Review by Rombur — APPROVED** [review]

File: containers/src/Kokkos_Bitset.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (crtrott) [reply containers/src/Kokkos_Bitset.hpp:98]: May need to do this with a "deprecated" guard.

**Comment by crtrott** [comment]
Three of the CUDA builds seg faulted during compilation. The CUDA 12 build was fine.

**Comment by dalg24** [comment]
Retest this please

**Pull request merged**